smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/TypeScriptSettings.java

```diff
@@ -313,4 +314,29 @@ public ShapeId resolveServiceProtocol(Model model, ServiceShape service, Set<Sha
     public void setProtocol(ShapeId protocol) {
         this.protocol = Objects.requireNonNull(protocol);
     }
+
+    /**
+     * An enum indicating the type of artifact the code generator will produce.
+     */
+    public enum ArtifactType {
+        CLIENT(SymbolVisitor::new),
+        SSDK((m, s) -> new ServerSymbolVisitor(m, new SymbolVisitor(m, s)));
+
+        private final BiFunction<Model, TypeScriptSettings, SymbolProvider> symbolProviderFactory;
+
+        ArtifactType(BiFunction<Model, TypeScriptSettings, SymbolProvider> symbolProviderFactory) {
+            this.symbolProviderFactory = symbolProviderFactory;
+        }
+
+        /**
+         * Creates a TypeScript symbol provider suited to the artifact type.
+         *
+         * @param model Model to generate symbols for.
+         * @param settings Settings used by the symbol provider.
+         * @return Returns the created provider.
+         */
+        public SymbolProvider createSymbolProvider(Model model, TypeScriptSettings settings) {
+            return symbolProviderFactory.apply(model, settings);
+        }
+    }
```

Comment:
This works, but the canonical, non-everything-has-to-be-a-function solution looks like this
```suggestion
public enum ArtifactType {
    CLIENT {
        @Override
        public SymbolProvider createSymbolProvider(Model model, TypeScriptSettings settings) {
            return new SymbolVisitor(model, settings);
        }
    },
    SSDK {
        @Override
        public SymbolProvider createSymbolProvider(Model model, TypeScriptSettings settings) {
            return new ServerSymbolVisitor(model, new SymbolVisitor(model, settings));
        }
    };

    /**
     * Creates a TypeScript symbol provider suited to the artifact type.
     *
     * @param model Model to generate symbols for.
     * @param settings Settings used by the symbol provider.
     * @return Returns the created provider.
     */
    public abstract SymbolProvider createSymbolProvider(Model model, TypeScriptSettings settings);
}
```